File: geoai-frontend/lib/plans.ts
export type PlanId = "guest" | "free" | "pro" | "premium" | "contract";

export type QuotaPeriod = "day" | "month";

export type PlanLimits = {
  label: string;
  priceMonthly: number | null;
  priceLabel: string;
  modules: ("gw" | "sw" | "lu")[];
  modes: string[];
  maxBBoxKm2: number | null;
  maxMonthsRange: number | null;
  analysesPerMonth: number;
  analysesPerDay?: number;
  quotaPeriod: QuotaPeriod;
  pdfExport: boolean;
  excelExport: boolean;
  whiteLabel: boolean;
  maxRegions: number | null;
  maxProvinces: number | null;
};

export const PLAN_LIMITS: Record<PlanId, PlanLimits> = {
  guest: {
    label: "Démo",
    priceMonthly: 0,
    priceLabel: "Sans compte",
    modules: ["lu"],
    modes: ["point", "region", "province"],
    maxBBoxKm2: 0,
    maxMonthsRange: 3,
    analysesPerMonth: 0,
    analysesPerDay: 999999,
    quotaPeriod: "day",
    pdfExport: false,
    excelExport: false,
    whiteLabel: false,
    maxRegions: 1,
    maxProvinces: 1,
  },
  free: {
    label: "Free",
    priceMonthly: 0,
    priceLabel: "Gratuit",
    modules: ["lu"],
    modes: ["point"],
    maxBBoxKm2: 0,
    maxMonthsRange: 6,
    analysesPerMonth: 10,
    quotaPeriod: "month",
    pdfExport: false,
    excelExport: false,
    whiteLabel: false,
    maxRegions: 1,
    maxProvinces: 1,
  },
  pro: {
    label: "Pro",
    priceMonthly: 39,
    priceLabel: "39 €/mois",
    modules: ["gw", "sw", "lu"],
    modes: ["point", "bbox", "province"],
    maxBBoxKm2: null,
    maxMonthsRange: 36,
    analysesPerMonth: 100,
    quotaPeriod: "month",
    pdfExport: true,
    excelExport: true,
    whiteLabel: false,
    maxRegions: null,
    maxProvinces: null,
  },
  premium: {
    label: "Premium",
    priceMonthly: 129,
    priceLabel: "129 €/mois",
    modules: ["gw", "sw", "lu"],
    modes: ["point", "bbox", "province", "region", "national"],
    maxMonthsRange: null,
    maxBBoxKm2: null,
    analysesPerMonth: 500,
    quotaPeriod: "month",
    pdfExport: true,
    excelExport: true,
    whiteLabel: true,
    maxRegions: null,
    maxProvinces: null,
  },
  contract: {
    label: "Contrat (B2B)",
    priceMonthly: null,
    priceLabel: "Sur devis",
    modules: ["gw", "sw", "lu"],
    modes: ["point", "bbox", "province", "region", "national"],
    maxMonthsRange: null,
    maxBBoxKm2: null,
    analysesPerMonth: 999999,
    quotaPeriod: "month",
    pdfExport: true,
    excelExport: true,
    whiteLabel: true,
    maxRegions: null,
    maxProvinces: null,
  },
};

export const PLAN_ORDER: PlanId[] = ["free", "pro", "premium", "contract"];

export function normalizePlan(plan: string | null | undefined): PlanId {
  if (plan === "guest" || plan === "pro" || plan === "premium" || plan === "contract") return plan;
  return "free";
}

export function isGuestPlan(plan: string | null | undefined): boolean {
  return plan === "guest";
}

export function getQuotaLimit(limits: PlanLimits): number {
  if (limits.quotaPeriod === "day" && limits.analysesPerDay != null) {
    return limits.analysesPerDay;
  }
  return limits.analysesPerMonth;
}

export function getQuotaLabel(plan: PlanId): string {
  const limits = PLAN_LIMITS[plan];
  if (limits.quotaPeriod === "day") {
    return `${limits.analysesPerDay ?? 0} / jour`;
  }
  return `${limits.analysesPerMonth} / mois`;
}

export function getLimits(plan: string | null | undefined): PlanLimits {
  return PLAN_LIMITS[normalizePlan(plan)];
}

export function monthsBetween(dateDebut: string, dateFin: string): number {
  const start = new Date(dateDebut);
  const end = new Date(dateFin);
  if (Number.isNaN(start.getTime()) || Number.isNaN(end.getTime())) return 0;
  return (
    (end.getFullYear() - start.getFullYear()) * 12 +
    (end.getMonth() - start.getMonth()) +
    1
  );
}

export function bboxAreaKm2(bbox: number[]): number | null {
  if (!bbox || bbox.length !== 4) return null;
  const [minLon, minLat, maxLon, maxLat] = bbox;
  const latMid = ((minLat + maxLat) / 2) * (Math.PI / 180);
  const widthKm = Math.abs(maxLon - minLon) * 111.32 * Math.cos(latMid);
  const heightKm = Math.abs(maxLat - minLat) * 110.574;
  return widthKm * heightKm;
}

export type PlanCheckInput = {
  plan: string | null | undefined;
  activeModule: string;
  activeMode: string;
  dateDebut: string;
  dateFin: string;
  bbox?: number[] | null;
  analysesUsed: number;
  zoneSelection?: any;
};

// Return keys instead of hardcoded strings
export function checkPlanAccess(input: PlanCheckInput): { ok: true } | { ok: false; messageKey: string; messageParams?: any } {
  const limits = getLimits(input.plan);

  if (!limits.modules.includes(input.activeModule as "gw" | "sw" | "lu")) {
    return {
      ok: false,
      messageKey: "ERR_MODULE_NOT_INCLUDED",
      messageParams: { plan: limits.label }
    };
  }

  if (!limits.modes.includes(input.activeMode)) {
    return {
      ok: false,
      messageKey: "ERR_MODE_NOT_INCLUDED",
      messageParams: { mode: input.activeMode, plan: limits.label }
    };
  }

  // Guest specific validation rules
  if (input.plan === "guest") {
    if (input.activeMode === "region" && input.zoneSelection?.code !== "Tadla-Azilal" && input.zoneSelection?.code !== "Tadla Azilal" && input.zoneSelection?.code !== "Tadla - Azilal") {
      return {
        ok: false,
        messageKey: "ERR_GUEST_REGION"
      };
    }
    if (input.activeMode === "province" && input.zoneSelection?.code !== "Béni Mellal") {
      return {
        ok: false,
        messageKey: "ERR_GUEST_PROVINCE"
      };
    }
    // Note: For 'point', the UI (Map.tsx) restricts clicks to 10 specific coordinates.
    // If backend validation is needed, add point coord checks here.
  }

  if (limits.maxMonthsRange != null) {
    const months = monthsBetween(input.dateDebut, input.dateFin);
    if (months > limits.maxMonthsRange) {
      return {
        ok: false,
        messageKey: "ERR_MAX_MONTHS",
        messageParams: { max: limits.maxMonthsRange, plan: limits.label }
      };
    }
  }

  if (input.activeMode === "bbox") {
    if (limits.maxBBoxKm2 === 0) {
      return {
        ok: false,
        messageKey: "ERR_BBOX_GUEST"
      };
    }
    if (limits.maxBBoxKm2 != null && input.bbox) {
      const area = bboxAreaKm2(input.bbox);
      if (area != null && area > limits.maxBBoxKm2) {
        return {
          ok: false,
          messageKey: "ERR_BBOX_TOO_LARGE",
          messageParams: { area: area.toFixed(1), max: limits.maxBBoxKm2 }
        };
      }
    }
  }

  const quotaLimit = getQuotaLimit(limits);
  if (input.analysesUsed >= quotaLimit) {
    return {
      ok: false,
      messageKey: input.plan === "guest" ? "ERR_QUOTA_GUEST" : "ERR_QUOTA_USER",
      messageParams: { limit: quotaLimit }
    };
  }

  return { ok: true };
}

export function currentUtcDay(): string {
  const d = new Date();
  return d.toISOString().slice(0, 10);
}

export function currentYearMonth(): string {
  const d = new Date();
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}`;
}

File: backend/config/plans.py
"""Plan limits — keep in sync with geoai-frontend/lib/plans.ts"""
from __future__ import annotations

import math
from datetime import datetime
from typing import Any

PlanId = str

PLAN_LIMITS: dict[str, dict[str, Any]] = {
    "guest": {
        "label": "Démo",
        "modules": ["lu"],
        "modes": ["point", "region", "province"],
        "max_bbox_km2": 0,
        "max_months_range": 3,
        "analyses_per_month": 0,
        "analyses_per_day": 999999,
        "quota_period": "day",
        "pdf_export": False,
        "excel_export": False,
        "max_regions": 1,
        "max_provinces": 1,
    },
    "free": {
        "label": "Free",
        "modules": ["lu"],
        "modes": ["point"],
        "max_bbox_km2": 0,
        "max_months_range": 6,
        "analyses_per_month": 10,
        "quota_period": "month",
        "pdf_export": False,
        "excel_export": False,
        "max_regions": 1,
        "max_provinces": 1,
    },
    "pro": {
        "label": "Pro",
        "modules": ["gw", "sw", "lu"],
        "modes": ["point", "bbox", "province"],
        "max_bbox_km2": None,
        "max_months_range": 36,
        "analyses_per_month": 100,
        "quota_period": "month",
        "pdf_export": True,
        "excel_export": True,
        "max_regions": None,
        "max_provinces": None,
    },
    "premium": {
        "label": "Premium",
        "modules": ["gw", "sw", "lu"],
        "modes": ["point", "bbox", "province", "region", "national"],
        "max_bbox_km2": None,
        "max_months_range": None,
        "analyses_per_month": 500,
        "quota_period": "month",
        "pdf_export": True,
        "excel_export": True,
        "max_regions": None,
        "max_provinces": None,
    },
    "contract": {
        "label": "Contrat (B2B)",
        "modules": ["gw", "sw", "lu"],
        "modes": ["point", "bbox", "province", "region", "national"],
        "max_bbox_km2": None,
        "max_months_range": None,
        "analyses_per_month": 999999,
        "quota_period": "month",
        "pdf_export": True,
        "excel_export": True,
        "max_regions": None,
        "max_provinces": None,
    },
}


def normalize_plan(plan: str | None) -> PlanId:
    if plan in ("guest", "pro", "premium", "contract"):
        return plan
    return "free"


def is_guest_principal(principal_id: str) -> bool:
    return principal_id.startswith("guest:")


def guest_uuid_from_principal(principal_id: str) -> str:
    return principal_id.removeprefix("guest:")


def get_limits(plan: str | None) -> dict[str, Any]:
    return PLAN_LIMITS[normalize_plan(plan)]


def get_quota_limit(limits: dict[str, Any]) -> int:
    if limits.get("quota_period") == "day":
        return int(limits.get("analyses_per_day") or 0)
    return int(limits["analyses_per_month"])


def months_between(date_debut: str, date_fin: str) -> int:
    start = datetime.strptime(date_debut, "%Y-%m-%d")
    end = datetime.strptime(date_fin, "%Y-%m-%d")
    return (end.year - start.year) * 12 + (end.month - start.month) + 1


def bbox_area_km2(bbox: list[float] | None) -> float | None:
    if not bbox or len(bbox) != 4:
        return None
    min_lon, min_lat, max_lon, max_lat = bbox
    lat_mid = math.radians((min_lat + max_lat) / 2)
    width_km = abs(max_lon - min_lon) * 111.32 * math.cos(lat_mid)
    height_km = abs(max_lat - min_lat) * 110.574
    return width_km * height_km


def current_year_month() -> str:
    now = datetime.utcnow()
    return f"{now.year}-{now.month:02d}"


def current_utc_day() -> str:
    return datetime.utcnow().strftime("%Y-%m-%d")


class PlanViolation(Exception):
    def __init__(self, message: str, status_code: int = 403):
        self.message = message
        self.status_code = status_code
        super().__init__(message)


def validate_plan_access(
    plan: str | None,
    *,
    active_module: str | None,
    granularite: str,
    date_debut: str | None = None,
    date_fin: str | None = None,
    bbox: list[float] | None,
    analyses_used: int,
    check_quota: bool = True,
) -> None:
    limits = get_limits(plan)

    if active_module and active_module not in limits["modules"]:
        raise PlanViolation(
            f"Module '{active_module}' not included in {limits['label']} plan.",
            403,
        )

    if granularite not in limits["modes"]:
        raise PlanViolation(
            f"Zone mode '{granularite}' requires a higher plan than {limits['label']}.",
            403,
        )

    max_months = limits["max_months_range"]
    if max_months is not None and date_debut and date_fin:
        try:
            if months_between(date_debut, date_fin) > max_months:
                raise PlanViolation(
                    f"Date range exceeds {max_months} months on {limits['label']} plan.",
                    403,
                )
        except ValueError as e:
            raise PlanViolation("Invalid date format. Use YYYY-MM-DD.", 400) from e

    if granularite == "bbox":
        max_bbox = limits["max_bbox_km2"]
        if max_bbox == 0:
            raise PlanViolation(
                "Bbox mode requires a free account. Sign up to continue.",
                403,
            )
        if max_bbox is not None and bbox:
            area = bbox_area_km2(bbox)
            if area is not None and area > max_bbox:
                raise PlanViolation(
                    f"Bbox area {area:.1f} km² exceeds {max_bbox} km² limit on Free plan.",
                    403,
                )

    quota_limit = get_quota_limit(limits)
    if check_quota and analyses_used >= quota_limit:
        period = "today" if limits.get("quota_period") == "day" else "this month"
        raise PlanViolation(
            f"Analysis quota reached ({quota_limit} {period}).",
            402,
        )
